File: 050-Pow-x-n/solution.py
class Solution(object):
    #修正前
    def myPow_original(self, x, n) -> float:
        result = 1
        half = abs(n) // 2
        if n < 0:
            for i in range(0, half):
                result *= 1 / x
        else:
            for i in range(0, half):
                result *= x
              
        return result * result * x if n & 1 else result * result
    #修正後
    def myPow(self, x, n) -> float:
        result = 1
        #負數n預處理
        if n < 0:
            x = 1 / x
            n = abs(n)
        print(f"init\nn = {n}, x = {x}, result = {result}")
        while n > 0:
            if n & 1:
                result *= x
            x *= x
            n //= 2
            print(f"n = {n}, x = {x}, result = {result}")
        return result
        
if __name__ == "__main__":
    x = 2.0
    n = 10
    print("result = ", Solution().myPow(x, n))

"""
快速冪影片教學
https://youtu.be/GbDtCFhq20A?si=tWEqyB_81Zd6tiRe
"""

File: 050-Pow-x-n/solution.go
package main

import (
	"fmt"
)

func myPow(x float64, n int) float64 {
	N := n
	if N < 0 {
		x = 1 / x
		N = -n
	}
	var result = 1.0
	fmt.Printf("init\nN = %d, x = %f, result = %f\n", N, x, result)
	for N > 0 {
		if N&1 == 1 {
			fmt.Println("N is odd")
			result *= x
		}
		x *= x
		N /= 2
		fmt.Printf("N = %d, x = %f, result = %f\n", N, x, result)
	}
	return result
}

func main() {
	x := 2.0
	n := 10
	ans := myPow(x, n)
	fmt.Println("ans = ", ans)
}

/*
快速冪影片教學
https://youtu.be/GbDtCFhq20A?si=tWEqyB_81Zd6tiRe
*/
